tests/test_ensemble.py

```diff
@@ -0,0 +1,101 @@
+"""
+Unit tests for PDF class
+"""
+import sys
+import os
+import numpy as np, scipy.stats as sps
+import unittest
+import qp
+
+from data import *
+
+
+class EnsembleTestCase(unittest.TestCase):
+
+    def setUp(self):
+        """
+        Make any objects that are used in multiple tests.
+        """
+        self.ens_n = build_ensemble(GEN_TEST_DATA['norm'])
+        
+    def tearDown(self):
+        "Clean up any mock data files created by the tests."
+        pass
+
+
+    def _run_ensemble_funcs(self, ens, xpts):
+        """Run the test for a practicular class"""
+
+        pdfs = ens.pdf(xpts)
+        cdfs = ens.cdf(xpts)
+        logpdfs = ens.logpdf(xpts)
+        logcdfs = ens.logcdf(xpts)
+
+        assert np.allclose(np.log(pdfs), logpdfs, atol=1e-9)
+        assert np.allclose(np.log(cdfs), logcdfs, atol=1e-9)
+                
+        binw = xpts[1:] - xpts[0:-1]
+        check_cdf = ((pdfs[:,0:-1] + pdfs[:,1:]) * binw /2).cumsum(axis=1) - cdfs[:,1:]
+        assert_all_small(check_cdf, atol=5e-2)
+
+        hist = ens.histogramize(xpts)[1]
+        hist_check = ens.frozen.histogramize(xpts)[1]
+        assert_all_small(hist-hist_check, atol=1e-5)
+
+        ppfs = ens.ppf(QUANTS)
+        check_ppf = ens.cdf(ppfs) - QUANTS
+        assert_all_small(check_ppf, atol=2e-2)
+
+        sfs = ens.sf(xpts)
+        check_sf = sfs + cdfs
+        assert_all_small(check_sf-1, atol=2e-2)
+        
+        isfs = ens.isf(QUANTS)
+        check_isf = ens.cdf(ppfs) + QUANTS[::-1]
+        assert_all_small(check_isf-1, atol=2e-2)
+
+        samples = ens.rvs(size=1000)
+        assert samples.shape[0] == ens.frozen.npdf
+        assert samples.shape[1] == 1000
+        
+        median = ens.median()
+        mean = ens.mean()
+        var = ens.var()
```

Comment:
var() and std() throw errors for older versions of scipy (1.3 and 1.4), so we should remember to put a requirement on scipy>=1.5.2 in the setup.py file

Reply:
Done